### DISCUSSION TITLE
How does Elk's AS2 content negotiation/discovery work?

### DISCUSSION THREAD

**Discussion by snarfed** [elk-zone/elk#2457] [Q&A]
Hi! I build and run another ActivityPub service, https://fed.brid.gy/ , and I'm trying to get it to interoperate with Elk, but I haven't figured out how to get Elk to serve AS2 data. When I fetch a profile or post URL with [AS2 content negotiation](https://www.w3.org/TR/activitypub/#retrieving-objects), eg `Accept: application/activity+json` or `Accept: application/ld+json`, I always get `Content-Type: text/html` returned. I've tried both signed (with HTTP Signatures) and unsigned fetches; no luck. I also don't see any `rel=alternate` links to AS2 URLs in the HTML or headers.

However, when I enter Elk profile/post URLs into Mastodon search, it finds them ok. Am I missing something obvious? Example curl command lines below. cc https://github.com/snarfed/bridgy-fed/issues/709. Thanks in advance!

```sh
$ curl -vL -H 'Accept: application/activity+json' https://elk.zone/m.webtoo.ls/@antfu
...
< HTTP/2 200 
< content-type: text/html;charset=utf-8
...

$ curl -vL -H 'Accept: application/ld+json' https://mozilla.social/@mconley/111348170686611324
...
< HTTP/2 200 
< content-type: text/html;charset=utf-8
...
``

**Comment by trwnh** [comment]
the web frontend has to forward requests based on conneg, and elk seems to only respond to `Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"`

**Reply by snarfed** [reply]
Interesting, thanks! You're right, mozilla.social at least does serve AS2 to that `Accept` header, but not any variations, eg my more complicated `Accept: application/activity+json; q=0.9, application/ld+json; profile="https://www.w3.org/ns/activitystreams"; q=0.8, text/html; charset=utf-8; q=0.7` gets `text/html` back, even though it includes that exact stanza.

I guess I'll file an issue to improve their conneg handling.

Other interesting minor points:
* Elk returns `Content-Type: application/activity+json; charset=utf-8`, ie _not_ `ld+json`, but it requires `ld+json` in the `Accept` header; it returns `text/html` to `Accept: application/activity+json; profile="https://www.w3.org/ns/activitystreams"`. Conneg, whee!
* elk.zone doesn't forward conneg requests at all, even with exactly `Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"`. I guess it doesn't have to, but that definitely hurts interop and AP/AS2 discovery. 🤷

**Reply by snarfed** [reply]
...and confirmed re elk.zone's AP discovery specifically. If you search for an elk.zone post URL on a Mastodon instance (logged in) that hasn't seen it before, eg https://elk.zone/m.webtoo.ls/@antfu/110045168476286703 on https://indieweb.social/ , you get no results. No AP/AS2 discovery happening there. Sigh.

**Reply by snarfed** [reply]
Filed https://github.com/elk-zone/elk/issues/2463
